Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs

```diff
@@ -91,34 +91,43 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
             var contestActivityEntity = await this.activityService
                 .GetContestActivity(contestDetailsServiceModel!.Map<ContestForActivityServiceModel>());
 
-            var participant = await this.participantsData
-                .GetWithProblemsForParticipantsByContestByUserAndIsOfficial(
-                    id,
-                    user.Id,
-                    contestActivityEntity.CanBeCompeted);
-
             contestDetailsServiceModel!.CanBeCompeted = contestActivityEntity.CanBeCompeted;
             contestDetailsServiceModel.CanBePracticed = contestActivityEntity.CanBePracticed;
-            contestDetailsServiceModel.IsAdminOrLecturerInContest = isLecturerOrAdmin;
+            contestDetailsServiceModel.IsAdminOrLecturerInContest = isLecturerInContestOrAdmin;
+
+            var userParticipants = await this.participantsData
+                .GetWithProblemsForParticipantsByContestByUser(id, user.Id)
+                .ToListAsync();
 
-            if (!isLecturerOrAdmin && participant != null && contestActivityEntity.CanBeCompeted)
+            var competeParticipant = userParticipants.FirstOrDefault(p => p.IsOfficial);
+
+            var participantToGetProblemsFrom = contestDetailsServiceModel.CanBeCompeted
+                ? competeParticipant
+                : userParticipants.FirstOrDefault(p => !p.IsOfficial);
```

Comment:
Can you extract `practiceParticiapnt` as well..
Also it's better to use `contestActivityEntity.CanBeCompeted` and set the `contestDetailsServiceModel` properties at the end of the method. This way you minimize the possibility for someone misusing some of the properties before being set and also it's more clear that the `contestActivityEntity.CanBeCompeted` is taking into account the user itself.

Reply:
Fixed